File: leetcode/0803-bricks-falling-when-hit/Solution.java
class Solution {
    private int[] parent, size;

    private int find(int x) {
        while (parent[x] != x) {
            parent[x] = parent[parent[x]];
            x = parent[x];
        }
        return x;
    }

    private void union(int a, int b) {
        int ra = find(a), rb = find(b);
        if (ra == rb) return;
        if (size[ra] < size[rb]) { int t = ra; ra = rb; rb = t; }
        parent[rb] = ra;
        size[ra] += size[rb];
    }

    public int[] hitBricks(int[][] grid, int[][] hits) {
        int rows = grid.length, cols = grid[0].length;
        int roof = rows * cols;
        int[][] directions = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};

        int[][] finalGrid = new int[rows][cols];
        for (int r = 0; r < rows; r++) finalGrid[r] = grid[r].clone();
        for (int[] hit : hits) finalGrid[hit[0]][hit[1]] = 0;

        parent = new int[rows * cols + 1];
        size = new int[rows * cols + 1];
        for (int i = 0; i < parent.length; i++) { parent[i] = i; size[i] = 1; }

        for (int r = 0; r < rows; r++) {
            for (int c = 0; c < cols; c++) {
                if (finalGrid[r][c] == 1) {
                    if (r == 0) union(idx(r, c, cols), roof);
                    for (int[] d : directions) {
                        int nr = r + d[0], nc = c + d[1];
                        if (nr >= 0 && nr < rows && nc >= 0 && nc < cols && finalGrid[nr][nc] == 1) {
                            union(idx(r, c, cols), idx(nr, nc, cols));
                        }
                    }
                }
            }
        }

        int[] result = new int[hits.length];
        for (int i = hits.length - 1; i >= 0; i--) {
            int r = hits[i][0], c = hits[i][1];
            if (grid[r][c] == 0) continue;

            int before = size[find(roof)];

            finalGrid[r][c] = 1;
            if (r == 0) union(idx(r, c, cols), roof);
            for (int[] d : directions) {
                int nr = r + d[0], nc = c + d[1];
                if (nr >= 0 && nr < rows && nc >= 0 && nc < cols && finalGrid[nr][nc] == 1) {
                    union(idx(r, c, cols), idx(nr, nc, cols));
                }
            }

            int after = size[find(roof)];
            result[i] = Math.max(0, after - before - 1);
        }

        return result;
    }

    private int idx(int r, int c, int cols) {
        return r * cols + c;
    }
}

File: leetcode/0803-bricks-falling-when-hit/solution.py
from typing import List


class UnionFind:
    def __init__(self, size: int):
        self.parent = list(range(size))
        self.size = [1] * size

    def find(self, x: int) -> int:
        while self.parent[x] != x:
            self.parent[x] = self.parent[self.parent[x]]
            x = self.parent[x]
        return x

    def union(self, a: int, b: int) -> None:
        ra, rb = self.find(a), self.find(b)
        if ra == rb:
            return
        # attach the smaller tree under the larger one, roll its size up
        if self.size[ra] < self.size[rb]:
            ra, rb = rb, ra
        self.parent[rb] = ra
        self.size[ra] += self.size[rb]

    def component_size(self, x: int) -> int:
        return self.size[self.find(x)]


class Solution:
    def hitBricks(self, grid: List[List[int]], hits: List[List[int]]) -> List[int]:
        rows, cols = len(grid), len(grid[0])
        roof = rows * cols  # virtual node representing "connected to the ceiling / row 0"

        def idx(r: int, c: int) -> int:
            return r * cols + c

        # step 1: compute the grid's final state after ALL hits have been applied
        final_grid = [row[:] for row in grid]
        for r, c in hits:
            final_grid[r][c] = 0

        # step 2: build union-find on that final state (the state after every hit)
        uf = UnionFind(rows * cols + 1)
        directions = [(-1, 0), (1, 0), (0, -1), (0, 1)]

        for r in range(rows):
            for c in range(cols):
                if final_grid[r][c] == 1:
                    if r == 0:
                        uf.union(idx(r, c), roof)
                    for dr, dc in directions:
                        nr, nc = r + dr, c + dc
                        if 0 <= nr < rows and 0 <= nc < cols and final_grid[nr][nc] == 1:
                            uf.union(idx(r, c), idx(nr, nc))

        # step 3: replay the hits in REVERSE, adding bricks back one at a time.
        # whatever newly becomes connected to the roof at each step is exactly
        # what would have fallen at that hit when time ran forward.
        result = [0] * len(hits)
        for i in range(len(hits) - 1, -1, -1):
            r, c = hits[i]
            if grid[r][c] == 0:
                continue  # there was never a brick here, nothing to report

            before = uf.component_size(roof)

            final_grid[r][c] = 1
            if r == 0:
                uf.union(idx(r, c), roof)
            for dr, dc in directions:
                nr, nc = r + dr, c + dc
                if 0 <= nr < rows and 0 <= nc < cols and final_grid[nr][nc] == 1:
                    uf.union(idx(r, c), idx(nr, nc))

            after = uf.component_size(roof)

            # subtract 1 for the brick we just placed back itself, it doesn't count as "fallen"
            fallen = after - before - 1
            result[i] = max(0, fallen)

        return result
